src/distributed/public_api.jl

```diff
@@ -0,0 +1,275 @@
+workers.
+"""
+function reduce!(
+        backend::AbstractLuxDistributedBackend, sendrecvbuf, op::F; root::Int=0) where {F}
+    return __reduce!(backend, sendrecvbuf, op, get_device(sendrecvbuf); root)
+end
+
+function reduce!(backend::AbstractLuxDistributedBackend,
+        sendbuf, recvbuf, op::F; root::Int=0) where {F}
+    send_dev = get_device(sendbuf)
+    recv_dev = get_device(recvbuf)
+    if send_dev === recv_dev
+        return __reduce!(backend, sendbuf, recvbuf, op, send_dev; root)
+    else
+        sendbuf_ = sendbuf |> recv_dev
+        @warn "`sendbuf` and `recvbuf` are on different devices." maxlog=1
+        __reduce!(backend, sendbuf_, recvbuf, op, recv_dev; root)
+        return recvbuf
+    end
+end
+
+function __reduce! end
+
+CRC.@non_differentiable reduce!(::Any...)
+
+# syncronize!
+"""
+    synchronize!!(backend::AbstractLuxDistributedBackend, ps; root::Int=0)
+
+Synchronize the given structure `ps` using the given backend. The value at `root` will be
+broadcasted to all other workers.
+"""
+function synchronize!!(backend::AbstractLuxDistributedBackend, ps::Tuple; root::Int=0)
+    length(ps) == 0 && return ps
+    return map(x -> synchronize!!(backend, x; root), ps)
+end
+
+function synchronize!!(backend::AbstractLuxDistributedBackend,
+        ps::NamedTuple{fields}; root::Int=0) where {fields}
+    length(ps) == 0 && return ps
+    return NamedTuple{fields}(map(x -> synchronize!!(backend, x; root), values(ps)))
+end
+
+function synchronize!!(
+        backend::AbstractLuxDistributedBackend, ps::AbstractArray{T}; root::Int=0) where {T}
+    if isbitstype(T)
+        bcast!(backend, ps; root)
+        return ps
+    end
+    return map(x -> synchronize!!(backend, x; root), ps)
+end
+
+function synchronize!!(backend::AbstractLuxDistributedBackend, ps::Leaf; root::Int=0)
+    @set! ps.state = synchronize!!(backend, ps.state; root)
+    return ps
+end
+
+function synchronize!!(backend::AbstractLuxDistributedBackend, ps::T; root::Int=0) where {T}
+    isbitstype(T) && return bcast!(backend, [ps]; root)[]
+    return ps # If we don't know how to synchronize, just return the value. For ex, Symbol, String, etc.
+end
+
+# data container
+"""
+    DistributedDataContainer(backend::AbstractLuxDistributedBackend, data)
+
+`data` must be compatible with `MLUtils` interface. The returned container is compatible
+with `MLUtils` interface and is used to partition the dataset across the available
+processes.
+"""
+@concrete struct DistributedDataContainer
+    data
+    idxs
+end
+
+function DistributedDataContainer(backend::AbstractLuxDistributedBackend, data)
+    total_size = length(data)
+    split_across = total_workers(backend)
+    size_per_worker = Int(ceil(total_size / split_across))
+
+    partitions = collect(Iterators.partition(1:total_size, size_per_worker))
+    idxs = collect(partitions[local_rank(backend) + 1])
+
+    return DistributedDataContainer(data, idxs)
+end
```

Comment:
`idxs` behavior doesn't seem right for `Tuples` which is an accepted data type for MLUtils.

For the example in https://github.com/LuxDL/Lux.jl/pull/500#issuecomment-1963180874

```julia
julia> DistributedUtils.DistributedDataContainer(backend, (x,y)).idxs
julia> DistributedUtils.DistributedDataContainer(backend, (x,y)).idxs
2-element Vector{Int64}:
 1
 2

julia> DistributedUtils.DistributedDataContainer(backend, x).idxs
256-element Vector{Int64}:
   1
   2
   3
   4
   5
   ⋮
 253
 254
 255
 256

julia> using MLUtils

julia> numobs(x)
256

julia> numobs((x,y))
256
```

Reply:
Could you elaborate on what the intended design should be? Are we expected to pass an `MLUtils.DataLoader` object or a `CuIterator` to DistributedDataContainer?

Reply:
@avik-pal ping

Reply:
This can be any container that supports getindex and length, but it wont recourse into objects. So in the case you showed, it needs to be `zip(x, y)`

Reply:
Now behavior is similar to MLUtils